File: Python/problem1_3.py
from fractions import gcd


def sum_multiples(a, b, end):
    def calc(m):
        n = (end - 1) // m
        return m * n * (n + 1) // 2
    return calc(a) + calc(b) - calc(a * b // gcd(a, b))


if __name__ == '__main__':
    print(sum_multiples(3, 5, 1000))

File: Scala/Problem1_3.scala
object Problem1_3 {
  def sumMultiples(a: BigInt)(b: BigInt)(end: BigInt): BigInt = {
    def calc(m: BigInt) = {
      val n = (end - 1) / m
      m * n * (n + 1) / 2
    }
    calc(a) + calc(b) - calc(a * b / (a gcd b))
  }

  def main(args: Array[String]): Unit =
    println(sumMultiples(3)(5)(1000))
}
